doc/notes/188_enhance_logging_ant.adoc

```diff
@@ -0,0 +1,245 @@
+some of these practices while we are changing to the `log4*` idiom of
+logging.
+
+The first Best Practice to logging is to author a mission statement for
+the logging and proceed from there.  This aligns with Shlaer-Mellor
+domain modelling quite nicely.
+
+== 3 Requirements
+
+. Establish a Project Munin logging mission statement to guide decision
+  making.
+. Make reporting flow from logging.
+. Support logging levels:
+  ** Fatal
+  ** Error
+  ** Warning
+  ** Information
+  ** Debug
+  ** Trace
+. Support conditional log message string arithmetic.
+. Establish conventions for enabled logging.
+. Support logger hierarchy.
+. Establish naming conventions for logger hierarchy.
+. Establish rules and guidelines for the construction of log messages.
+. Facilitate machine parsing of log messages (using structured logging).
+  TODO: Consider the format that Prometheus supports.
+. Identify tooling that will be used to harvest the log data.
+. Protect privacy and security by logging application-level information
+  judiciously.
+
+== 4 Analysis
+
+=== 4.1 Munin Logging Mission Statement
+
+* Munin software shall log information critical to the verification of protocols.
+  This includes detection of alarms, failures and successes.
+* Munin software shall be instrumented to enable the measurement of the
+  performance of the Protocol Verifier application itself (benchmarking).
+* Munin software shall be instrumented sparingly with debug statements
+  adequate for defect isolation (debugging).
+* The volume of production-level logging shall not exceed the volume of
+  production audit event traffic.
+
+=== 4.2 Breakdown of Application Traffic
+
+The Protocol Verifier monitors devices through telemetry.  There are two
+levels of detail in this telemetry, that of a _job_ and that of an
+_event_.  In general, the user is interested in the status of a job.
+Events are interesting only when they are erroneous.  Summaries of
+aggratated events are interesting in terms of throughput.
+
+=== 4.3 Job Logging
+
+The primary deliverable log entry is the final status of a job.  A job
+completes with status _successful_, _failed_ or _alarm_.  Together with
+the job type (name) and identifier, this is the key status provided by the
+Protocol Verifier.  Failed and alarmed jobs must also report internal
+state satisfactory to understanding failure modes in application-level
+jobs in the system being monitored.  Canonical logging practices shall be
+applied to facilitate problem determination for failed and alarmed jobs.
+
+To provide canonical log messages once and only once per job, information
+about the job and events will be accumulated in instance attributes of the
+modelled artefacts.  At the end of the job, the accumulated information
+will be structured and delivered to the logger.
+
+==== 4.3.1 Job-Level Statistics for Successful Jobs
+
+For successful jobs, the following shall be reported:
+
+* job type (name)
+* job identifier
+* event count
+
```

Comment:
Should this be supplemented with an indication of whether the job was a retry or not (and if, so, the job id of which job it was retrying)?

Reply:
That is a great idea.  But let us add it when we (if) add retry detection.